src/gen9_hevc_encoder.c

```diff
@@ -0,0 +1,7361 @@
+    i965_free_gpe_resource(&priv_ctx->res_brc_pak_statistic_buffer);
+    i965_free_gpe_resource(&priv_ctx->res_brc_input_buffer_for_enc_kernels);
+    i965_free_gpe_resource(&priv_ctx->res_brc_constant_data_buffer);
+
+    i965_free_gpe_resource(&priv_ctx->res_mb_code_surface);
+
+    // free VME buffers
+    i965_free_gpe_resource(&priv_ctx->res_flatness_check_surface);
+    i965_free_gpe_resource(&priv_ctx->res_brc_me_dist_buffer);
+    i965_free_gpe_resource(&priv_ctx->s4x_memv_distortion_buffer);
+    i965_free_gpe_resource(&priv_ctx->s4x_memv_data_buffer);
+    i965_free_gpe_resource(&priv_ctx->s16x_memv_data_buffer);
+    i965_free_gpe_resource(&priv_ctx->s32x_memv_data_buffer);
+    i965_free_gpe_resource(&priv_ctx->res_32x32_pu_output_buffer);
+    i965_free_gpe_resource(&priv_ctx->res_simplest_intra_buffer);
+    i965_free_gpe_resource(&priv_ctx->res_kernel_debug);
+    i965_free_gpe_resource(&priv_ctx->res_sad_16x16_pu_buffer);
+    i965_free_gpe_resource(&priv_ctx->res_vme_8x8_mode_buffer);
+    i965_free_gpe_resource(&priv_ctx->res_intra_mode_buffer);
+    i965_free_gpe_resource(&priv_ctx->res_intra_distortion_buffer);
+    i965_free_gpe_resource(&priv_ctx->res_vme_uni_sic_buffer);
+    i965_free_gpe_resource(&priv_ctx->res_con_corrent_thread_buffer);
+    i965_free_gpe_resource(&priv_ctx->res_mv_index_buffer);
+    i965_free_gpe_resource(&priv_ctx->res_mvp_index_buffer);
+    i965_free_gpe_resource(&priv_ctx->res_roi_buffer);
+    i965_free_gpe_resource(&priv_ctx->res_mb_statistics_buffer);
+
+    if (priv_ctx->scaled_2x_surface_obj) {
+          i965_DestroySurfaces(priv_ctx->ctx, &priv_ctx->scaled_2x_surface_id, 1);
+          priv_ctx->scaled_2x_surface_obj = NULL;
+    }
+
+    // free PAK buffers
+    i965_free_gpe_resource(&priv_ctx->deblocking_filter_line_buffer);
+    i965_free_gpe_resource(&priv_ctx->deblocking_filter_tile_line_buffer);
+    i965_free_gpe_resource(&priv_ctx->deblocking_filter_tile_column_buffer);
+    i965_free_gpe_resource(&priv_ctx->metadata_line_buffer);
+    i965_free_gpe_resource(&priv_ctx->metadata_tile_line_buffer);
+    i965_free_gpe_resource(&priv_ctx->metadata_tile_column_buffer);
+    i965_free_gpe_resource(&priv_ctx->sao_line_buffer);
+    i965_free_gpe_resource(&priv_ctx->sao_tile_line_buffer);
+    i965_free_gpe_resource(&priv_ctx->sao_tile_column_buffer);
+
+    priv_ctx->res_inited = 0;
+}
+
+#define ALLOC_GPE_RESOURCE(RES, NAME, SIZE)                 \
+    do{                                                     \
+        i965_free_gpe_resource(&priv_ctx->RES);             \
+        if (!i965_allocate_gpe_resource(i965->intel.bufmgr, \
+                                 &priv_ctx->RES,            \
+                                 SIZE,                      \
+                                 NAME))                     \
+            goto FAIL;                                      \
+        i965_zero_gpe_resource(&priv_ctx->RES);             \
```

Comment:
it is only called once in the initialization, not for each frame encoding. so it impacts the performance really a little